tools/pubsys/src/aws/validate_ami/results.rs

```diff
@@ -0,0 +1,714 @@
+            }),
+        }
+    }
+}
+
+/// Represents a single EC2 image validation result
+#[derive(Debug, Eq, Hash, PartialEq, Serialize)]
+pub(crate) struct AmiValidationResult {
+    /// The id of the image
+    pub(crate) image_id: String,
+
+    /// The expected value of the image
+    pub(crate) expected_value: ImageDef,
+
+    /// The actual retrieved value of the image
+    pub(crate) actual_value: Option<ImageDef>,
+
+    /// The region the image resides in
+    #[serde(serialize_with = "serialize_region")]
+    pub(crate) region: Region,
+
+    /// The validation status of the image
+    pub(crate) status: AmiValidationResultStatus,
+}
+
+fn serialize_region<S>(region: &Region, serializer: S) -> std::result::Result<S::Ok, S::Error>
+where
+    S: serde::Serializer,
+{
+    serializer.serialize_str(region.to_string().as_str())
+}
+
+impl AmiValidationResult {
+    pub(crate) fn new(
+        image_id: String,
+        expected_value: ImageDef,
+        actual_value: Option<ImageDef>,
+        region: Region,
+    ) -> Self {
+        // Determine the validation status based on equality, presence, and absence of expected and
+        // actual image values
+        let status = match (&expected_value, &actual_value) {
+            (expected_value, Some(actual_value)) if actual_value.eq(expected_value) => {
+                AmiValidationResultStatus::Correct
+            }
+            (_, Some(_)) => AmiValidationResultStatus::Incorrect,
+            (_, None) => AmiValidationResultStatus::Missing,
+        };
+        AmiValidationResult {
+            image_id,
+            expected_value,
+            actual_value,
+            region,
+            status,
+        }
+    }
+}
+
+#[derive(Tabled, Serialize)]
+struct AmiValidationRegionSummary {
+    correct: i32,
+    incorrect: i32,
+    missing: i32,
+    accessible: bool,
+}
+
+impl From<&HashSet<AmiValidationResult>> for AmiValidationRegionSummary {
+    fn from(results: &HashSet<AmiValidationResult>) -> Self {
+        let mut region_validation = AmiValidationRegionSummary {
+            correct: 0,
+            incorrect: 0,
+            missing: 0,
+            accessible: true,
+        };
+        for validation_result in results {
+            match validation_result.status {
+                AmiValidationResultStatus::Correct => region_validation.correct += 1,
+                AmiValidationResultStatus::Incorrect => region_validation.incorrect += 1,
+                AmiValidationResultStatus::Missing => region_validation.missing += 1,
+            }
+        }
+        region_validation
+    }
+}
+
+impl AmiValidationRegionSummary {
+    fn no_valid_results() -> Self {
+        // When the images in a region couldn't be retrieved, use `-1` to indicate this in the output table
+        // and set `accessible` to `false`
+        AmiValidationRegionSummary {
+            correct: -1,
+            incorrect: -1,
+            missing: -1,
+            accessible: false,
```

Comment:
It seems like this boolean might be duplicating the information that -1, -1, -1 represents. Not a big deal though.

Reply:
It does, but I wanted an extra column that clearly indicates there was a problem with the region itself instead of just vague `-1` cells.